### PR TITLE
Add command to remove stale review features

### PR DESCRIPTION
**Pull request by phillebaba**

### CODE REVIEW

File: action.yaml [hunk 1]
```diff
@@ -26,7 +26,7 @@ inputs:
 runs:
   using: docker
   #image: docker://ghcr.io/xenitab/gitops-promotion:v1.0.0-pre2
-  image: docker://phillebaba/gitops-promotion:branch-deploy
+  image: docker://phillebaba/gitops-promotion:stale-feature
```
Comment (simongottschlag) [inline action.yaml:29]: Is this correct?

File: pkg/command/command.go [hunk 2]
```diff
@@ -92,7 +93,9 @@ func Run(ctx context.Context, args []string) (string, error) {
 		if err != nil {
 			return "", err
 		}
-		return FeatureCommand(ctx, cfg, repo, *group, *app, *tag, *feature)
+		return FeatureNewCommand(ctx, cfg, repo, *group, *app, *tag, *feature)
+  case "feature-stale":
+    return FeatureDeleteStaleCommand(ctx, cfg, repo, 7*24*time.Hour)
```
Comment (simongottschlag) [inline pkg/command/command.go:98]: Can we make the duration a flag to begin with?

File: pkg/command/feature.go [hunk 3]
```diff
@@ -59,7 +67,85 @@ func FeatureCommand(ctx context.Context, cfg config.Config, repo *git.Repository
...
+  }
+
+  // Remove feature directories that have not been committed to for longer than max age
+  for _, state := range states {
+    commit, err := repo.GetLastCommitForPath(state.AppPath())
+    if err != nil {
+      return "", err
+    }
+    if time.Now().Sub(commit.Author().When) < maxAge {
+      continue
```
Comment (simongottschlag) [inline pkg/command/feature.go:119]: Add a log here to make it clear why it's skipped

File: pkg/command/feature.go [hunk 4]
```diff
@@ -59,7 +67,85 @@ func FeatureCommand(ctx context.Context, cfg config.Config, repo *git.Repository
...
+
+  // Commit, push branch, create PR
+  branchName := "remove/stale-feature"
+  err := repo.CreateBranch(branchName, true)
+	if err != nil {
+		return "", fmt.Errorf("could not create branch: %w", err)
+	}
+  title := "Remove stale review features"
+  description := ""
+	sha, err := repo.CreateCommit(branchName, title)
```
Comment (simongottschlag) [inline pkg/command/feature.go:135]: Seems like we are missing go fmt on this file

File: pkg/command/feature.go [hunk 5]
```diff
@@ -59,7 +67,85 @@ func FeatureCommand(ctx context.Context, cfg config.Config, repo *git.Repository
...
+		return "", fmt.Errorf("could not commit changes: %w", err)
+	}
+	err = repo.Push(branchName, true)
+	if err != nil {
+		return "", fmt.Errorf("could not push changes: %w", err)
+	}
+	/*auto, err := cfg.IsEnvironmentAutomated(environmentName)
+	if err != nil {
+		return "", fmt.Errorf("could not get environment automation state: %w", err)
+	}*/
```
Comment (simongottschlag) [inline pkg/command/feature.go:146]: Can we remove this?

File: pkg/git/git.go [hunk 6]
```diff
@@ -166,6 +165,76 @@ func (g *Repository) CreateCommit(branchName, message string) (*git2go.Oid, erro
 	return sha, nil
 }
 
+// GetLastCommitForPath returns the last commit for the given path. All files and subdirectories
+// will be considered if the path is a directory.
+func (g *Repository) GetLastCommitForPath(path string) (*git2go. Commit, error) {
```
Comment (simongottschlag) [inline pkg/git/git.go:170]: Damn!

**Review by simongottschlag — COMMENTED** [review]
Looking good Mr! I've left a few comments from browsing through the PR. 😊👍

**Comment by bittrance** [comment]
I'm worried that the strategy of having a single PR for all cleanup is going to be hard to merge. I can see scenarios where a sequence of "still waiting for X" will mean that the cleanup PR will always need manual edits. Unfortunately, I think we need one PR per feature.

File: pkg/command/feature.go [hunk 7]
```diff
@@ -3,18 +3,21 @@ package command
...
+	"time"
 
 	"github.com/spf13/afero"
+
 	"github.com/xenitab/gitops-promotion/pkg/config"
 	"github.com/xenitab/gitops-promotion/pkg/git"
 	"github.com/xenitab/gitops-promotion/pkg/manifest"
 )
 
 // FeatureCommand is similar to NewCommand but creates a PR with a temporary deployment of the application.
```
Comment (NissesSenap) [inline pkg/command/feature.go:18]: I think it should say `FeatureNewCommand` in the comment. Surprised that the linter don't say anything about this.

File: pkg/command/feature.go [hunk 8]
```diff
@@ -59,7 +67,91 @@ func FeatureCommand(ctx context.Context, cfg config.Config, repo *git.Repository
...
+	if !removedApplication {
+		return "No stale application to remove, exiting early.", nil
+	}
+
+	// Commit, push branch, create PR
+	branchName := "remove/stale-feature"
+	err := repo.CreateBranch(branchName, true)
+	if err != nil {
+		return "", fmt.Errorf("could not create branch: %w", err)
+	}
```
Comment (NissesSenap) [inline pkg/command/feature.go:139]: What happens if this branch already exists?

File: pkg/command/feature.go [hunk 9]
```diff
@@ -59,7 +67,91 @@ func FeatureCommand(ctx context.Context, cfg config.Config, repo *git.Repository
...
+	description := ""
+	sha, err := repo.CreateCommit(branchName, title)
+	if err != nil {
+		return "", fmt.Errorf("could not commit changes: %w", err)
+	}
+	err = repo.Push(branchName, true)
+	if err != nil {
+		return "", fmt.Errorf("could not push changes: %w", err)
+	}
+	auto, err := cfg.IsEnvironmentAutomated(environmentName)
```
Comment (NissesSenap) [inline pkg/command/feature.go:150]: Can a feature branch exist in a none dev env?

File: pkg/manifest/manifest.go [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package manifest
+
+func RemoveStaleFeature() {
+
+}
```
Comment (NissesSenap) [inline pkg/manifest/manifest.go:5]: Should probably be removed

**Review by NissesSenap — COMMENTED** [review]
A few minor comments.

File: action.yaml [hunk 11]
```diff
@@ -25,8 +25,7 @@ inputs:
     required: false
 runs:
   using: docker
-  #image: docker://ghcr.io/xenitab/gitops-promotion:v1.0.0-pre2
-  image: docker://phillebaba/gitops-promotion:branch-deploy
+  image: docker://ghcr.io/xenitab/gitops-promotion:v1.0.0
```
[the remark this answers is not in the document]
Comment (phillebaba) [reply action.yaml:28]: Yes I will create a new PR before release with the correct image tag

**Review by NissesSenap — APPROVED** [review]

**Comment by phillebaba** [comment]
I have disabled e2e tests for now as they need a large refactor. Will get this done later in the summer.

**Pull request merged**